### PR TITLE
common/storage: layer & sqlite3 implementation

### PR DESCRIPTION
**Pull request by tsvisabo**
Initial sketching of storage layer - for review

### CODE REVIEW

File: common/errors.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#ifndef COMMON_ERRORS_H_
+#define COMMON_ERRORS_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+enum retcode_t {
+    E_OK = 0,
+    E_UNKNOWN,
```
Comment (th0br0) [inline common/errors.h]: We should use a category/subcategory/severity approach here?

File: common/storage/persistable.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#include "common/trinary/trit_array.h"
+#include "common/errors.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct persistable_base persistable_t;
+
+typedef struct {
```
Comment (th0br0) [inline common/storage/persistable.h]: not fully following here. could you explain your approach?

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: common/storage/persistable.h (hunk 2 above)
Comment (tsvisabo) [reply common/storage/persistable.h]: Every view model should extend `persistable_t`
the base struct holds a ptr to the opaque ptr of the derived struct
so when you call `stor_store/stor_load` you just need to send the base ptr and you get the key/value of the underlying object

Comment (tsvisabo) [reply common/storage/persistable.h]: Also, i'm not exactly sure about `table_name`, but some table identifier needs to be included

Comment (th0br0) [reply common/storage/persistable.h]: How about we make this not so generic? 
It's perfectly fine to limit the available 'tables' via a enum in code and then just have a single method for storing/loading everything imho.
The `vtable` adds significant overhead.

Comment (tsvisabo) [reply common/storage/persistable.h]: Alright, that might be simpler

File: common/storage/sql/schema.sql [hunk 3]
```diff
@@ -0,0 +1,17 @@
+CREATE TABLE IF NOT EXISTS iota_transaction (
+  signature_or_message BLOB NOT NULL,
+  address BLOB NOT NULL,
+  value INTEGER DEFAULT 0 NOT NULL,
+  obsolete_tag BLOB,
+  timestamp TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
```
Comment (th0br0) [inline common/storage/sql/schema.sql]: should not default.

File: common/storage/sql/schema.sql [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  current_index SMALLINT DEFAULT 0 NOT NULL,
```
Comment (th0br0) [inline common/storage/sql/schema.sql]: should not default

File: common/storage/sql/schema.sql [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  last_index SMALLINT DEFAULT 0 NOT NULL,
```
Comment (th0br0) [inline common/storage/sql/schema.sql]: should not default. these are important for essence

File: common/storage/sql/schema.sql [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  bundle BLOB NOT NULL,
+  trunk BLOB NOT NULL,
+  branch BLOB NOT NULL,
+  tag BLOB NOT NULL,
+  attachment_timestamp TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
```
Comment (th0br0) [inline common/storage/sql/schema.sql]: why default?

File: common/storage/sql/sqlite3/connection.cc [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  int rc = sqlite3_exec((sqlite3*)conn->db, statement, 0, 0, &errMsg);
+  if (rc != SQLITE_OK) {
+    // TODO - log
+    sqlite3_free(errMsg);
+    return RC_SQLITE3_FAILED_CREATE_INDEX_DB;
+  }
+
+  //TODO - implement connections pool so no two threads
```
Comment (th0br0) [inline common/storage/sql/sqlite3/connection.cc]: maybe contemplate 

```
  conn->execute("PRAGMA journal_mode = WAL");
  conn->execute("PRAGMA foreign_keys = ON");
  conn->execute("PRAGMA busy_timeout = 10000");
```

File: common/storage/sql/sqlite3/connection.cc [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    // TODO - log
+    // LOG(FATAL) << __FUNCTION__ << " Can't open database: \n"
+    //          << sqlite3_errmsg((sqlite3*)conn->db);
+  } else {
+    // TODO - log
+    // LOG(INFO) << "Opened database successfully";
+  }
+
+  if (config->indexApprovee) {
+    if (retcode = create_index(conn, "trunk_index", "trunk")) {
```
Comment (th0br0) [inline common/storage/sql/sqlite3/connection.cc]: won't these fail if the indices exist already?

File: common/storage/sql/sqlite3/connection.cc [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+retcode_t create_index(const connection_t* const conn,
+                       const char* const indexName, const char* const colName) {
+  char* errMsg = 0;
+
+  char statement[MAX_CREATE_INDEX_STATEMENT_SIZE];
+
+  int res = snprintf(statement, MAX_CREATE_INDEX_STATEMENT_SIZE,
+                     "CREATE INDEX IF NOT EXISTS %s ON %s(%s)",
+                     TRANSACTION_TABLE_NAME, indexName, colName);
```
Comment (th0br0) [inline common/storage/sql/sqlite3/connection.cc]: let's keep this a bit more generic and make `TRANSACTION_TABLE_NAME` a parameter?

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: common/storage/sql/schema.sql (hunk 6 above)
Comment (tsvisabo) [reply common/storage/sql/schema.sql]: right, will fix it, should it be nullable though?

File: common/storage/sql/sqlite3/connection.cc (hunk 8 above)
Comment (tsvisabo) [reply common/storage/sql/sqlite3/connection.cc]: it creates if index does not exist already

File: common/storage/sql/sqlite3/storage.cc [hunk 10]
```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.cc]: this file should be called `storage.c`

File: common/storage/sql/statements.cc [hunk 11]
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+#include <stdio.h>
```
Comment (th0br0) [inline common/storage/sql/statements.cc]: this file should be called `statements.c`

File: common/storage/sql/sqlite3/storage.cc [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    logger_helper_print(SQLITE3_LOGGER_ID, LOGGER_ERR, "Failed in statement",
+                        statement);
+    sqlite3_free(err_msg);
+    return RC_SQLITE3_FAILED_INSERT_DB;
+  }
+
+  return RC_OK;
+}
+
+retcode_t iota_stor_update(const connection_t* const conn,
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.cc]: add a FIXME or TODo

**Review by th0br0 — CHANGES_REQUESTED** [review]
what happened to metadata storage?

File: common/storage/sql/sqlite3/connection.cc [hunk 13]
```diff
@@ -0,0 +1,142 @@
+/*
```
Comment (th0br0) [inline common/storage/sql/sqlite3/connection.cc]: should be called `conenction.c`

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: common/logger_helper.h [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+ * Refer to the LICENSE file for licensing information
+ */
+
+#ifndef COMMON_LOGGER_HELPER_H_
+#define COMMON_LOGGER_HELPER_H_
+
+#include <logger.h>
+#include <stdarg.h>
+#include <stdbool.h>
+#include "common/lock_handle.h"
```
Comment (th0br0) [inline common/logger_helper.h]: unused include

**Review by th0br0 — CHANGES_REQUESTED** [review]
remove `common/storage/sql/ciri.db`

what about tests?

File: common/storage/connection.h [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+extern "C" {
+#endif
+
+typedef struct connection_t {
+  void* db;
+} connection_t;
+
+typedef struct connection_config_t {
+  const char* dbPath;
+  bool indexAddress;
```
Comment (thibault-martinez) [inline common/storage/connection.h]: Maybe don't mix camelCase and snake_case ?

File: common/logger_helper.h [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#ifndef COMMON_LOGGER_HELPER_H_
+#define COMMON_LOGGER_HELPER_H_
+
+#include <logger.h>
+#include <stdarg.h>
```
Comment (thibault-martinez) [inline common/logger_helper.h]: Not sure this include is needed here

File: common/storage/main.cc [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+ */
+
+#include <stdio.h>
+
+#include <gflags/gflags.h>
+#include <logger.h>
+
+#include "common/model/transaction.h"
+#include "common/storage/storage.h"
+#include "connection.h"
```
Comment (thibault-martinez) [inline common/storage/main.cc]: absolute path

File: common/logger_helper.h [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct logger_t logger_t;
+
+extern void logger_helper_init(const char* const logger_id,
+                               logger_level_t level, bool enable_color);
+extern void logger_helper_destroy(const char* const logger_id);
+extern void logger_helper_print(const char* const logger_id,
```
Comment (thibault-martinez) [inline common/logger_helper.h:23]: See https://github.com/iotaledger/entangled/issues/77#issuecomment-408060855

File: common/storage/sql/sqlite3/connection.c [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+                        "Failed to open db on path: %s", config->dbPath);
+  } else {
+    logger_helper_print(CONNECTION_LOGGER_ID, LOGGER_INFO,
+                        "Opened database (from path: %s ) successfully",
+                        config->dbPath);
+  }
+
+  if (config->indexApprovee) {
+    if (retcode = create_index_if_not_exist(conn, TRANSACTION_TABLE_NAME,
+                                            "trunk_index", "trunk")) {
```
Comment (thibault-martinez) [inline common/storage/sql/sqlite3/connection.c]: Don't we have/want these in sql/defs.h ?

File: common/storage/sql/sqlite3/storage.c [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+    logger_helper_print(SQLITE3_LOGGER_ID, LOGGER_ERR,
+                        "Failed to retrieve columns");
+    return RC_SQLITE3_FAILED_SELECT_DB;
+  }
+
+  pack = (iota_transactions_pack*)txs;
+  if (pack->num_loaded < pack->num_txs) {
+    iota_transaction_t curr_transaction = pack->txs[pack->num_loaded++];
+
+    memcpy(curr_transaction->signature_or_message, col_values[c++], 2187);
```
Comment (thibault-martinez) [inline common/storage/sql/sqlite3/storage.c:51]: Don't we have these in constants ?

File: common/storage/connection.h (hunk 15 above)
Comment (tsvisabo) [reply common/storage/connection.h]: sure

File: common/logger_helper.h (hunk 16 above)
Comment (tsvisabo) [reply common/logger_helper.h]: true

File: common/storage/sql/sqlite3/storage.c (hunk 20 above)
Comment (tsvisabo) [reply common/storage/sql/sqlite3/storage.c:51]: Nope, and it's temporary until  #2 is merged and then i can use the size field in the trit_array

File: common/errors.h [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+#define COMMON_ERRORS_H_
+
+#include <stdio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file  erros.h
```
Comment (th0br0) [inline common/errors.h]: typo

File: common/helpers/files.c [hunk 22]
```diff
@@ -0,0 +1,57 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "common/helpers/files.h"
+
+int cp(const char *to, const char *from) {
```
Comment (th0br0) [inline common/helpers/files.c]: would we give this a better name please? `copy_file` or so?

File: common/storage/sql/sqlite3/connection.c [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+#include "common/storage/sql/defs.h"
+
+#include <sqlite3.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+#define CONNECTION_LOGGER_ID "sqlite3_connection_storage"
+
+retcode_t create_index_if_not_exist(const connection_t* const conn,
```
Comment (th0br0) [inline common/storage/sql/sqlite3/connection.c]: `if_not_exists`

File: common/storage/sql/sqlite3/connection.c [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+#include "common/storage/connection.h"
+#include "common/logger_helper.h"
+#include "common/storage/sql/defs.h"
+
+#include <sqlite3.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+#define CONNECTION_LOGGER_ID "sqlite3_connection_storage"
```
Comment (th0br0) [inline common/storage/sql/sqlite3/connection.c]: could we name this `stor_sqlite3_conn` (so in descending order in the dependency tree, not ascending)

File: common/storage/sql/sqlite3/storage.c [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+
+#include "common/logger_helper.h"
+#include "common/storage/sql/defs.h"
+#include "common/storage/sql/statements.h"
+#include "common/storage/storage.h"
+
+#include <logger.h>
+#include <sqlite3.h>
+
+#define SQLITE3_LOGGER_ID "sqlite3_storage"
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: `stor_sqlite3`?

File: common/storage/sql/sqlite3/storage.c (hunk 20 above)
Comment (th0br0) [reply common/storage/sql/sqlite3/storage.c:51]: create an issue to track this please?

File: common/storage/sql/sqlite3/storage.c [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+                        (void*)&pack, &err_msg);
+  *num_loaded = pack.num_loaded;
+
+  if (rc != SQLITE_OK) {
+    logger_helper_print(SQLITE3_LOGGER_ID, LOGGER_ERR, "Failed in statement",
+                        statement);
+    sqlite3_free(err_msg);
+    return RC_SQLITE3_FAILED_INSERT_DB;
+  }
+
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: remove whitelines

File: common/storage/sql/statements.c [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+                        statement);
+    return RC_SQL_FAILED_WRITE_STATEMENT;
+  }
+}
+
+retcode_t iota_transactions_update_statement(const char* index_col,
+                                             const trit_array_p key,
+                                             const iota_transaction_t tx,
+                                             char statement[],
+                                             size_t statement_cap) {}
```
Comment (th0br0) [inline common/storage/sql/statements.c:126]: add some log output `not implemented` or similar?
(& create todo issue?)

**Review by th0br0 — CHANGES_REQUESTED** [review]
just nits

**Review by th0br0 — APPROVED** [review]

**Pull request merged**